crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,771 @@
+use bevy_platform_support::{
+    prelude::Vec,
+    sync::{
+        atomic::{AtomicBool, AtomicPtr, AtomicU32, AtomicU64, Ordering},
+        Arc, Weak,
+    },
+};
+use core::{mem::ManuallyDrop, num::NonZero};
+use log::warn;
+
+use crate::query::DebugCheckedUnwrap;
+
+use super::{Entity, EntitySetIterator};
+
+/// This is the item we store in the free list.
+struct Slot {
+    entity_index: AtomicU32,
+    entity_generation: AtomicU32,
+}
+
+impl Slot {
+    // TODO: could maybe make this `&mut`??
+    #[inline]
+    fn set_entity(&self, entity: Entity) {
+        self.entity_generation
+            .store(entity.generation(), Ordering::Relaxed);
+        self.entity_index.store(entity.index(), Ordering::Relaxed);
+    }
+
+    #[inline]
+    fn get_entity(&self) -> Entity {
+        Entity {
+            index: self.entity_index.load(Ordering::Relaxed),
+            // SAFETY: This is not 0 since it was from an entity's generation.
+            generation: unsafe {
+                NonZero::new_unchecked(self.entity_generation.load(Ordering::Relaxed))
+            },
+        }
+    }
+}
+
+/// Each chunk stores a buffer of [`Slot`]s at a fixed capacity.
+struct Chunk {
+    /// Points to the first slot. If this is null, we need to allocate it.
+    first: AtomicPtr<Slot>,
+}
+
+impl Chunk {
+    const NUM_CHUNKS: u32 = 24;
+    const NUM_SKIPPED: u32 = u32::BITS - Self::NUM_CHUNKS;
+
+    /// Computes the capacity of the chunk at this index within [`Self::NUM_CHUNKS`].
+    /// The first 2 have length 512 (2^9) and the last has length (2^31)
+    #[inline]
+    fn capacity_of_chunk(chunk_index: u32) -> u32 {
+        // We do this because we're skipping the first 8 powers, so we need to make up for them by doubling the first index.
+        // This is why the first 2 indices both have a capacity of 256.
+        let corrected = chunk_index.max(1);
+        // We add NUM_SKIPPED because the total capacity should be as if [`Self::NUM_CHUNKS`] were 32.
+        // This skips the first NUM_SKIPPED powers.
+        let corrected = corrected + Self::NUM_SKIPPED;
+        // This bit shift is just 2^corrected.
+        1 << corrected
+    }
+
+    /// For this index in the whole buffer, returns the index of the [`Chunk`] and the index within that chunk.
+    #[inline]
+    fn get_indices(full_idnex: u32) -> (u32, u32) {
+        // We're countint leading zeros since each chunk has power of 2 capacity.
+        // So the leading zeros will be proportional to the chunk index.
+        let leading = full_idnex
+            .leading_zeros()
+            // We do a min because we skip the first 8 powers.
```

Comment:
Could we expand on this? Perhaps, "We do a min because we skip the first 8 powers to make space for the first chunk's entity count."